omigami/feature_evaluator.py

```diff
@@ -1,105 +1,54 @@
+        X_train = evaluation_data.train_data.X
+        y_train = evaluation_data.train_data.y
+        X_test = evaluation_data.test_data.X
+        y_test = evaluation_data.test_data.y
 
-        X_test = self._X[test_idx, :][:, features]
-        y_test = self._y[test_idx]
+        estimator = self._estimator.clone().fit(X_train, y_train)
         y_pred = estimator.predict(X_test)
 
         score = -self._metric(y_test, y_pred)
-        feature_ranks = self._get_feature_ranks(estimator, features)
-        return FeatureEvaluationResults(test_score=score, ranks=feature_ranks)
-
-    def get_n_features(self):
-        return self._n_features
-
-    def get_inner_loop_size(self):
-        return self._n_inner
-
-    def _make_metric(self, metric: Union[str, MetricFunction]) -> MetricFunction:
-        """Build metric function using the input `metric`. If a metric is a string
-        then is interpreted as a scikit-learn metric score, such as "accuracy".
-        Else, if should be a callable accepting two input arrays."""
-        if isinstance(metric, str):
-            return self._make_metric_from_string(metric)
-        elif hasattr(metric, "__call__"):
-            return metric
-        else:
-            raise ValueError("Input metric is not valid")
-
-    @staticmethod
-    def _make_metric_from_string(metric_string: str) -> MetricFunction:
-        if metric_string == "MISS":
-            return miss_score
-        if metric_string in SCORERS:
-            # pylint: disable=protected-access
-            return get_scorer(metric_string)._score_func
-        raise ValueError("Input metric is not a valid string")
-
-    def _get_feature_importances(self, estimator: Estimator):
-        if hasattr(estimator, "feature_importances_"):
-            return estimator.feature_importances_
-        elif hasattr(estimator, "coef_"):
-            return np.abs(estimator.coef_[0])
-        elif hasattr(estimator, "steps"):
-            for _, step in estimator.steps:
-                if hasattr(step, "coef_") or hasattr(step, "feature_importances_"):
-                    return self._get_feature_importances(step)
-        else:
-            raise ValueError("The estimator provided has no feature importances")
+        ranks = self._get_feature_ranks(estimator, features)
+        return FeatureEvaluationResults(test_score=score, ranks=ranks, model=estimator)
 
     def _get_feature_ranks(
-        self, estimator: Estimator, features: Iterable[int]
+        self, estimator: Estimator, features: Union[List[int], NumpyArray]
     ) -> FeatureRanks:
-        """Extract the feature rank from the input estimator. So far it can only handle
-        estimators as scikit-learn ones, so either having the `feature_importances_` or
-        the `coef_` attribute."""
-        feature_importances = self._get_feature_importances(estimator)
+        if self._n_initial_features == 0:
+            raise ValueError("Call set_n_initial_features first")
```

Comment:
Wouldn't it be more time saving if this check was at beginning of evaluate_features method?

Reply:
:+1: